### PR TITLE
Rewrite and test Github auth logic, separating standard and search quota

### PR DESCRIPTION
**Pull request by paulmelnikow**
Summary:

- Create new priority queue-based token cycling logic which correctly handles simultaneous requests using the same token
- Correctly support search API rate limits, which are separate from core API rate limits (#1119)
- When tokens are exhausted, return an error instead of attempting fallback to our API key
- Use a static token in development, when one is configured (replacing #1118)
- Split up Github auth logic into separate modules which can be tested in isolation

Todo:

- [x] Test persistence
- [x] Integration test?
- [x] Incorporate recent changes to `lib/github-auth.js`
- [x] Catch persistence errors and send them to sentry
- [x] Rework TokenProvider / TokenPool abstraction which may not be so helpful. Github has two pools so things feel a little bit off.
- [x] Update token persistence for these changes to the api provider
- [x] Wait for merge of #2400

To handle in follow on work:

- [ ] Log / emit to sentry when tokens are exhausted
- [x] (maybe) Consolidate reserve logic, probably in `Token`

### CODE REVIEW

File: lib/github-auth/acceptor.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
+'use strict';
+
+const request = require('request');
+const queryString = require('query-string');
+const serverSecrets = require('../server-secrets');
+const log = require('../log');
+
```
Comment (paulmelnikow) [inline lib/github-auth/acceptor.js]: This code has no tests so it was imported as is.

**Comment by paulmelnikow** [comment]
I'm eager to get this shipped, though I'm hesitant to do it until I get access to logs…

File: lib/github-auth/acceptor.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+      };
+      request(options, function(err, res, body) {
+        if (err != null) { return reject(err); }
+        resolve();
+      });
+    });
+  }));
+}
+
+function constEq(a, b) {
```
Comment (LVMBDV) [inline lib/github-auth/acceptor.js]: I'm new to node in general but why not use [crypto.timingSafeEqual(a, b)](https://nodejs.org/api/crypto.html#crypto_crypto_timingsafeequal_a_b) instead?

Comment (paulmelnikow) [reply lib/github-auth/acceptor.js]: I didn't write this, just moved it. Yes, happy to switch that over. It seems like that solves the same problem.

**Comment by paulmelnikow** [comment]
Since we have access to errors through Sentry, I think I will revive this and see if we can't nudge it forward.

**Comment by paulmelnikow** [comment]
As I keep working with this, I'm remembering how much work it was to understand the old code, to rework the code to make it testable, and to write the tests. I'm pretty happy with most of this, though I did add a checklist at the top of a few things I'd like to change.

Would someone be willing to pair with me to review, test, and merge this code? It would be great to have a second person's help.

We've merged a little bit of this work in #1736 but there quite a bit left. At this point it seems like a more manageable way forward would be through a series of small PRs, picking this in bit by bit.

**Comment by paulmelnikow** [comment]
After merging #1812, these could be the next steps:

1. Pull in the reworked initialization and persistence code
2. Move the token acceptor routes
3. Deal with the token and auth logic itself

**Comment by paulmelnikow** [comment]
This pull request **introduces 1 alert** when merging 2ffc96517246a1987353f1240f230c42494fe458 into e46a6fbde13e58562a2868381399ae23fc3f21d6 - [view on LGTM.com](https://lgtm.com/projects/g/badges/shields/rev/pr-d397eeefe629f14b2e9733299f3628cae0c0dd26)

**new alerts:**

* 1 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by paulmelnikow** [comment]
Have labeled this a blocker, as it affects the work of #1838.

**Comment by paulmelnikow** [comment]
This pull request **introduces 1 alert** when merging e334cc81958960733ccb704bcb3d35ced485e04c into a524658e74791a4fe98e4f058a3bb11ec2fa837f - [view on LGTM.com](https://lgtm.com/projects/g/badges/shields/rev/pr-6d9db9846cdddfe8d79089339680fc13a4b2769c)

**new alerts:**

* 1 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by paulmelnikow** [comment]
This pull request **introduces 3 alerts** when merging eb372a92a360c63ff9ed611b4f732613245f9530 into a524658e74791a4fe98e4f058a3bb11ec2fa837f - [view on LGTM.com](https://lgtm.com/projects/g/badges/shields/rev/pr-22de24a4d4693c7c4cd3f0a19ae2deedf331a914)

**new alerts:**

* 3 for Property access on null or undefined

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by paulmelnikow** [comment]
This feels like the home stretch. I updated the todos in the top comment.

**Comment by paulmelnikow** [comment]
This pull request **introduces 3 alerts** when merging 4bcc87654d64334ede17ab172fdefcdb9673aefa into a524658e74791a4fe98e4f058a3bb11ec2fa837f - [view on LGTM.com](https://lgtm.com/projects/g/badges/shields/rev/pr-3f040cfdba1feb83e013be4d4f60b847ae0bf478)

**new alerts:**

* 3 for Property access on null or undefined

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by paulmelnikow** [comment]
This pull request **introduces 3 alerts** when merging 19f398cd74f0988aec549da0d0b73f40c937e769 into a524658e74791a4fe98e4f058a3bb11ec2fa837f - [view on LGTM.com](https://lgtm.com/projects/g/badges/shields/rev/pr-44d5a9eed926da2ac41a293ef9750aab2e7c9270)

**new alerts:**

* 3 for Property access on null or undefined

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by paulmelnikow** [comment]
This pull request **introduces 3 alerts** when merging 375d60923fe697ea91bafe5a7fae2c3f77de79f1 into a524658e74791a4fe98e4f058a3bb11ec2fa837f - [view on LGTM.com](https://lgtm.com/projects/g/badges/shields/rev/pr-e576433ac04807b4a1b71909d3933b9fcf1f1f31)

**new alerts:**

* 3 for Property access on null or undefined

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by paulmelnikow** [comment]
This pull request **introduces 3 alerts** when merging 7ea08cc61a60f65f0085a3ddbfe8f2945d703e61 into 59fdd8a5b0512592f142eb4a1b76dba11bd8e0c9 - [view on LGTM.com](https://lgtm.com/projects/g/badges/shields/rev/pr-dde502451d6eccd6325233088e4bdc270d3a1ee8)

**new alerts:**

* 3 for Property access on null or undefined

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by paulmelnikow** [comment]
This pull request **introduces 3 alerts** when merging 7ea08cc61a60f65f0085a3ddbfe8f2945d703e61 into 38ca6fc396e101594ac624577d40a803817022c3 - [view on LGTM.com](https://lgtm.com/projects/g/badges/shields/rev/pr-8ea506d96d470145be6e886ba15d9debda3545f3)

**new alerts:**

* 3 for Property access on null or undefined

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by paulmelnikow** [comment]
This pull request **introduces 3 alerts** when merging e37d076a18e190675ed7d1df6891588600644d9d into d97a5e46978f96126eb5a786bd60df78401872e9 - [view on LGTM.com](https://lgtm.com/projects/g/badges/shields/rev/pr-6bc9ac0197f9b6431b97a82780793a2ef7762f29)

**new alerts:**

* 3 for Property access on null or undefined

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by paulmelnikow** [comment]
This pull request **introduces 1 alert** when merging d6258e448555c66be149753de3fd00620345211b into d97a5e46978f96126eb5a786bd60df78401872e9 - [view on LGTM.com](https://lgtm.com/projects/g/badges/shields/rev/pr-680cdc9cc7f30d38f3d84daf6da9eca87b0a9ffe)

**new alerts:**

* 1 for Unused variable, import, function or class

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: lib/fs-token-persistence.js [hunk 3]
```diff
@@ -23,21 +22,28 @@ class FsTokenPersistence extends TokenPersistence {
     }
 
     const tokens = JSON.parse(contents)
-    tokens.forEach(tokenString => {
-      githubAuth.addGithubToken(tokenString)
-    })
```
Comment (paulmelnikow) [inline lib/fs-token-persistence.js]: Now the caller's responsibility.

File: lib/redis-token-persistence.js [hunk 4]
```diff
@@ -21,16 +20,8 @@ class RedisTokenPersistence extends TokenPersistence {
...
-    let tokens
-    try {
-      tokens = await lrange(this.key, 0, -1)
-    } catch (e) {
-      log.error(e)
-    }
-
-    tokens.forEach(tokenString => {
-      githubAuth.addGithubToken(tokenString)
-    })
```
Comment (paulmelnikow) [inline lib/redis-token-persistence.js]: Now the caller's responsibility.

File: services/github/auth/acceptor.spec.js [hunk 5]
```diff
@@ -43,7 +43,10 @@ describe('Github token acceptor', function() {
   })
```
Comment (calebcartwright) [inline services/github/auth/acceptor.spec.js]: I couldn't add this above, but refers to top level describe block on line 14:

Can't we use sinon here via:
```js
before(function() {
    // make sure properties exist
    serverSecrets['gh_client_id'] = undefined
    serverSecrets['shields_ips'] = undefined
    sinon.stub(serverSecrets, 'gh_client_id').value(fakeClientId)
    ...
  })
```

I'm not sure if that buys us anything, but we've done the same elsewhere in other services (for example https://github.com/badges/shields/blob/master/services/jira/jira-test-helpers.js#L10)

File: services/github/auth/acceptor.spec.js [hunk 6]
```diff
@@ -43,7 +43,10 @@ describe('Github token acceptor', function() {
...
 
   beforeEach(function() {
-    acceptor.setRoutes(camp)
+    acceptor.setRoutes({
+      server: camp,
+      onTokenAccepted: () => {},
```
Comment (calebcartwright) [inline services/github/auth/acceptor.spec.js]: If I'm following the code flow correctly, this is what was extracted from the persistence implementations with the responsibility shifting to the caller. If that's right then should we attach a spy to this function to make sure it is called with the expected value?

File: services/github/github-api-provider.spec.js [hunk 7]
```diff
@@ -1,14 +1,50 @@
 'use strict'
 
 const { expect } = require('chai')
+const sinon = require('sinon')
```
Comment (calebcartwright) [inline services/github/github-api-provider.spec.js]: 👍 on all the tests in here

**Comment by calebcartwright** [comment]
I've made it through my first pass and added some inline comments. Although beware, there are more questions for my own understanding than there is useful feedback for you 😁

File: lib/fs-token-persistence.spec.js [hunk 8]
```diff
@@ -41,18 +37,17 @@ describe('File system token persistence', function() {
...
     context('when tokens are added', function() {
       it('saves the change', async function() {
         const newToken = 'e'.repeat(40)
-        const expected = initialTokens.slice()
+        const expected = Array.from(initialTokens)
         expected.push(newToken)
 
         await persistence.initialize()
-        githubAuth.addGithubToken(newToken)
         await persistence.noteTokenAdded(newToken)
```
[the remark this answers is not in the document]
Comment (paulmelnikow) [reply lib/fs-token-persistence.spec.js]: Should we change it to `add()`?

File: services/github/github-api-provider.js [hunk 9]
```diff
@@ -1,15 +1,79 @@
...
+  }
+
+  serializeDebugInfo({ sanitize = true } = {}) {
+    if (this.withPooling) {
+      return {
+        standardTokens: this.standardTokens.serializeDebugInfo({ sanitize }),
+        searchTokens: this.searchTokens.serializeDebugInfo({ sanitize }),
+      }
+    } else {
+      return {}
```
[the remark this answers is not in the document]
Comment (paulmelnikow) [reply services/github/github-api-provider.js]: When debug logging is turned on, it gets called on a timer in GithubConstellation.

File: lib/fs-token-persistence.spec.js [hunk 10]
```diff
@@ -41,18 +37,17 @@ describe('File system token persistence', function() {
     })
```
[the remark this answers is not in the document]
Comment (calebcartwright) [reply lib/fs-token-persistence.spec.js]: Yeah this test is the most critical for validating that core piece of functionality. It doesn't feel like a unit test (at least not in the pure sense), but I don't think the label matters all that much here

File: lib/fs-token-persistence.spec.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (calebcartwright) [reply lib/fs-token-persistence.spec.js]: Yeah I'd vote for `add()` or `addToken()` (and definitely `add()` if we can/will add things other than tokens)

File: services/github/auth/acceptor.spec.js (hunk 6 above)
Comment (paulmelnikow) [reply services/github/auth/acceptor.spec.js]: Yes, that makes sense. There was not yet a test of this part of the flow so I added one.

**Comment by paulmelnikow** [comment]
> Although beware, there are more questions for my own understanding than

Avoiding knowledge bottlenecks is fully half of why we do this :)

**Comment by calebcartwright** [comment]
I think this is good to go, will do another pass here in a bit and approve

File: services/github/auth/acceptor.spec.js [hunk 11]
```diff
@@ -112,4 +117,16 @@ describe('Github token acceptor', function() {
...
+
+  it('should add a received token', async function() {
+    const fakeAccessToken = 'its-my-token'
+
+    const { body } = await got(`${baseUrl}/github-auth/add-token`, {
+      form: true,
+      body: { shieldsSecret: fakeShieldsSecret, token: fakeAccessToken },
+    })
+
+    expect(onTokenAccepted).to.have.been.calledWith(fakeAccessToken)
```
Comment (calebcartwright) [inline services/github/auth/acceptor.spec.js]: 👍

**Comment by paulmelnikow** [comment]
Thanks so much! Will be glad to close this out… and see if it _really_ works. Let’s one of us tackle the renames in a follow-on.

**Comment by paulmelnikow** [comment]
I'll merge and deploy this in the afternoon when we get past peak usage time, in case anything breaks.

**Pull request merged**